fem/libceed/ceed.cpp

```diff
@@ -110,21 +110,97 @@ void InitCeedCoeff(Coefficient *Q, Mesh &mesh,
 #endif
 }
 
-void CeedPAAssemble(const CeedPAOperator& op,
-                    CeedData& ceedData)
+void InitCeedVecCoeff(VectorCoefficient *VQ, Mesh &mesh,
+                      const IntegrationRule &ir, CeedData *ptr)
+{
+#ifdef MFEM_USE_CEED
+   if (VectorConstantCoefficient *coeff =
+          dynamic_cast<VectorConstantCoefficient*>(VQ))
+   {
+      const int vdim = coeff->GetVDim();
+      const Vector &val = coeff->GetVec();
+      CeedVecConstCoeff *ceedCoeff = new CeedVecConstCoeff;
+      for (int i = 0; i < vdim; i++)
+      {
+         ceedCoeff->val[i] = val[i];
+      }
+      ptr->coeff_type = CeedCoeff::VecConst;
+      ptr->coeff = static_cast<void*>(ceedCoeff);
+   }
+   else if (VectorGridFunctionCoefficient* coeff =
+               dynamic_cast<VectorGridFunctionCoefficient*>(VQ))
+   {
+      CeedVecGridCoeff *ceedCoeff = new CeedVecGridCoeff;
+      ceedCoeff->coeff = coeff->GetGridFunction();
+      InitCeedVector(*ceedCoeff->coeff, ceedCoeff->coeffVector);
+      ptr->coeff_type = CeedCoeff::VecGrid;
+      ptr->coeff = static_cast<void*>(ceedCoeff);
+   }
+   else if (VectorQuadratureFunctionCoefficient *cQ =
+               dynamic_cast<VectorQuadratureFunctionCoefficient*>(VQ))
+   {
+      CeedVecQuadCoeff *ceedCoeff = new CeedVecQuadCoeff;
+      const int ne = mesh.GetNE();
+      const int dim = mesh.Dimension();
+      const int nq = ir.GetNPoints();
+      const QuadratureFunction &qFun = cQ->GetQuadFunction();
+      MFEM_VERIFY(qFun.Size() == dim * nq * ne,
+                  "Incompatible QuadratureFunction dimension \n");
+
+      MFEM_VERIFY(&ir == &qFun.GetSpace()->GetElementIntRule(0),
+                  "IntegrationRule used within integrator and in"
+                  " QuadratureFunction appear to be different");
+      qFun.Read();
+      ceedCoeff->coeff.MakeRef(const_cast<QuadratureFunction &>(qFun),0);
+      InitCeedVector(ceedCoeff->coeff, ceedCoeff->coeffVector);
+      ptr->coeff_type = CeedCoeff::VecQuad;
+      ptr->coeff = static_cast<void*>(ceedCoeff);
+   }
+   else
+   {
+      CeedVecQuadCoeff *ceedCoeff = new CeedVecQuadCoeff;
+      const int ne = mesh.GetNE();
+      const int dim = mesh.Dimension();
+      const int nq = ir.GetNPoints();
+      ceedCoeff->coeff.SetSize(dim * nq * ne);
+      auto C = Reshape(ceedCoeff->coeff.HostWrite(), dim, nq, ne);
+      DenseMatrix Q_ir;
+      for (int e = 0; e < ne; ++e)
```

Comment:
How expensive is this loop in general? To me it looks like linear cost in the number of quad points, so I almost want to just do this all the time regardless of the type of `VQ`. That would make the code simpler (remove all the `select...case` blocks), but I don't know if it is worth the performance penalty.

Reply:
I think this loop is much more expensive than the other cases in general (it is never computed on GPU for instance).